nameserver/zone_lookup.go

```diff
@@ -0,0 +1,301 @@
+			zone.mx.Unlock()
+			Warning.Printf("[zonedb] '%s' insertion for %s failed: %s", answer, target, err)
+			return nil, err
+		}
+	}
+	zone.mx.Unlock()
+	return
+}
+
+// Perform a lookup for a name in the zone
+// The name can be resolved locally with the local database or with some other resolution method (eg, a mDNS query)
+func (zone *zoneDb) DomainLookupName(name string) (res []ZoneRecord, err error) {
+	name = dns.Fqdn(name)
+	Debug.Printf("[zonedb] Looking for name '%s' in local(&remote) database", name)
+
+	zone.mx.RLock()
+	now := zone.clock.Now()
+	for identName, nameset := range zone.idents {
+		for _, ze := range nameset.getEntriesForName(name) {
+			// filter the entries with expired TTL
+			// locally introduced entries are never expired: they always have TTL=0
+			if ze.hasExpired(now) {
+				Debug.Printf("[zonedb] '%s': expired entry '%s' ignored: removing", name, ze)
+				nameset.deleteName(name)
+			} else {
+				res = append(res, ze)
+			}
+		}
+		if identName != defaultRemoteIdent {
+			nameset.touchName(name, now)
+		}
+	}
+	zone.mx.RUnlock()
+
+	if len(res) > 0 {
+		Debug.Printf("[zonedb] '%s' resolved in local database", name)
+	} else {
+		res, err = zone.domainLookup(name, zone.mdnsCli.LookupName)
+	}
+
+	if len(res) > 0 {
+		zone.startUpdatingName(name)
+		return res, nil
+	}
+	return nil, LookupError(name)
+}
+
+// Perform a lookup for a IP address in the zone
+// The address can be resolved either with the local database or
+// with some other resolution method (eg, a mDNS query)
+func (zone *zoneDb) DomainLookupInaddr(inaddr string) (res []ZoneRecord, err error) {
+	revIPv4, err := raddrToIPv4(inaddr)
+	if err != nil {
+		return nil, newParseError("lookup address", inaddr)
+	}
+
+	Debug.Printf("[zonedb] Looking for address in local(&remote) database: '%s' (%s)", revIPv4, inaddr)
+
+	zone.mx.RLock()
+	now := zone.clock.Now()
+	for identName, nameset := range zone.idents {
+		for _, ze := range nameset.getEntriesForIP(revIPv4) {
+			// filter the entries with expired TTL
+			// locally introduced entries are never expired: they always have TTL=0
+			if ze.hasExpired(now) {
+				Debug.Printf("[zonedb] '%s': expired entry '%s' ignored: removing", revIPv4, ze)
+				nameset.deleteIP(revIPv4)
+			} else {
+				res = append(res, ZoneRecord(ze))
+				if identName != defaultRemoteIdent {
+					nameset.touchName(ze.Name(), now)
+				}
+			}
+		}
+	}
+	zone.mx.RUnlock()
+
+	if len(res) > 0 {
+		Debug.Printf("[zonedb] '%s' resolved in local database", inaddr)
+	} else {
+		res, err = zone.domainLookup(inaddr, zone.mdnsCli.LookupInaddr)
+	}
+
+	if len(res) > 0 {
+		// note: even for reverse addresses, we perform the background updates in the name, not in the IP
+		//       this simplifies the process and produces basically the same results...
+		// note: we do not spend time trying to update names that did not return an initial response...
+		for _, r := range res {
+			zone.startUpdatingName(r.Name())
+		}
+
+		return res, nil
```

Comment:
Again, we should make responses unique at this point. Otherwise, `['ident1', 'name1', IP1]` and `['ident2', 'name1', IP1]` would produce `['name1', 'name1']`